pkg/resourcemanager/cosmosdb/sqldatabase/cosmosdbsql.go

```diff
@@ -0,0 +1,275 @@
+	}
+
+	if lroPollResult == pollclient.PollResultTryAgainLater {
+		// Need to wait a bit before trying again
+		return false, nil
+	}
+	if lroPollResult == pollclient.PollResultBadRequest {
+		// Reached a terminal state
+		instance.Status.SetFailedProvisioning(instance.Status.Message)
+		return true, nil
+	}
+
+	notFoundErrorCodes := []string{
+		errhelp.ParentNotFoundErrorCode,
+		errhelp.ResourceGroupNotFoundErrorCode,
+		errhelp.NotFoundErrorCode,
+		errhelp.ResourceNotFound,
+	}
+
+	// Get the existing resource
+	sqlDatabase, err := cosmosDBSQLDatabaseClient.GetSQLDatabase(
+		ctx,
+		instance.Spec.ResourceGroup,
+		instance.Spec.CosmosDBAccount,
+		instance.ObjectMeta.Name)
+	if err != nil {
+		azerr := errhelp.NewAzureError(err)
+		if azerr.Type != errhelp.ResourceNotFound && azerr.Type != errhelp.NotFoundErrorCode {
+			instance.Status.Message = fmt.Sprintf("failed GET-ing CosmosDBSQLDatabase: %s", err.Error())
+			return errhelp.IsErrorFatal(err, notFoundErrorCodes, nil)
+		}
+	}
+
+	// TODO: There's a bug in Cosmos DB SQL Database where the throughput settings configured on
+	// TODO: the DB don't actually get shown on the DB, instead they are shown on the throughput child resource.
+	// TODO: We monitor that resource too so we can diff against it
+	sqlDatabaseThroughputSettings, err := cosmosDBSQLDatabaseClient.GetSQLDatabaseThroughput(
+		ctx,
+		instance.Spec.ResourceGroup,
+		instance.Spec.CosmosDBAccount,
+		instance.ObjectMeta.Name)
+	if err != nil {
+		azerr := errhelp.NewAzureError(err)
+		if azerr.Type != errhelp.ResourceNotFound && azerr.Type != errhelp.NotFoundErrorCode {
+			instance.Status.Message = fmt.Sprintf("failed GET-ing CosmosDBSQLDatabaseThroughput: %s", err.Error())
+			return errhelp.IsErrorFatal(err, notFoundErrorCodes, nil)
+		}
+	}
+
+	// Transform our spec to Azure shape
+	transformed := m.transformToCosmosDBSQLDatabase(instance)
+
+	// Check if we're in the goal state already
+	resourceMatchesAzure := DoesResourceMatchAzure(transformed, sqlDatabase, sqlDatabaseThroughputSettings)
+	if resourceMatchesAzure {
+		instance.Status.SetProvisioned(resourcemanager.SuccessMsg)
+		instance.Status.ResourceId = *sqlDatabase.ID
+		return true, nil
+	}
+
+	// Drive to goal state
+	instance.Status.SetProvisioning("")
+	future, err := cosmosDBSQLDatabaseClient.CreateUpdateSQLDatabase(
+		ctx,
+		instance.Spec.ResourceGroup,
+		instance.Spec.CosmosDBAccount,
+		instance.ObjectMeta.Name,
+		transformed)
+	if err != nil {
+		instance.Status.Message = err.Error()
+		allowedErrors := []string{
+			errhelp.AsyncOpIncompleteError,
+			errhelp.AlreadyExists,
+			errhelp.FailoverGroupBusy,
+		}
+		unrecoverableErrors := []string{
+			errhelp.InvalidFailoverGroupRegion,
+		}
+		return errhelp.IsErrorFatal(err, append(allowedErrors, notFoundErrorCodes...), unrecoverableErrors)
+	}
+	instance.Status.SetProvisioning("Resource request successfully submitted to Azure")
+	instance.Status.PollingURL = future.PollingURL()
```

Comment:
Is there anywhere that clears the polling url once the long-running operation completes? Should there be?

Reply:
It's done in the polling helper (which is awkward and side-effecty but is how it's set up right now)